### PR TITLE
TLSv1.3: Send supported_groups in EE

### PR DESCRIPTION
**Pull request by mattcaswell**
The TLSv1.3 spec says that a server SHOULD send supported_groups in the EE message if there is a group that it prefers to the one used in the key_share. Clients MAY act on that. At the moment we don't do anything with it on the client side, but that may change in the future.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: ssl/statem/extensions_srvr.c [hunk 1]
```diff
@@ -868,6 +868,63 @@ int tls_construct_stoc_ec_pt_formats(SSL *s, WPACKET *pkt, unsigned int context,
...
+
+    /* Copy group ID if supported */
+    for (i = 0; i < numgroups; i++, groups += 2) {
+        if (tls_curve_allowed(s, groups, SSL_SECOP_CURVE_SUPPORTED)) {
+            if (first) {
+                /*
+                 * Check if the client is already using our preferred group. If
+                 * so we don't need to add this extension
+                 */
+                if (s->s3->group_id == GET_GROUP_ID(groups, 0))
```
Comment (kaduk) [inline ssl/statem/extensions_srvr.c]: Hmm, perhaps this call would be more clear with 'i' than '0'.  But maybe not.

File: ssl/statem/extensions_srvr.c [hunk 2]
```diff
@@ -868,6 +868,63 @@ int tls_construct_stoc_ec_pt_formats(SSL *s, WPACKET *pkt, unsigned int context,
...
+                        || !WPACKET_start_sub_packet_u16(pkt)) {
+                    SSLerr(SSL_F_TLS_CONSTRUCT_STOC_SUPPORTED_GROUPS,
+                           ERR_R_INTERNAL_ERROR);
+                    return 0;
+                }
+
+                first = 0;
+            }
+            if (!WPACKET_put_bytes_u8(pkt, groups[0])
+                || !WPACKET_put_bytes_u8(pkt, groups[1])) {
```
Comment (kaduk) [inline ssl/statem/extensions_srvr.c]: Do we really need two WPACKET calls here?  Wouldn't a memcpy_u16 work fine?  Or even something involving GET_GROUP_ID().

File: ssl/statem/extensions_srvr.c (hunk 1 above)
Comment (mattcaswell) [inline ssl/statem/extensions_srvr.c]: That would require changing the loop to not add 2 to "groups" every time - which has other implications that probably make it less clear that way overall.

File: ssl/statem/extensions_srvr.c (hunk 2 above)
Comment (mattcaswell) [inline ssl/statem/extensions_srvr.c]: There isn't a `WPACKET_memcpy_u16()` only a `WPACKET_memcpy()` which requires an explicit length. Perhaps you are confusing it with `WPACKET_sub_memcpy_u16()`. I changed this to WPACKET_put_bytes_u16() with GET_GROUP_ID().

**Comment by mattcaswell** [comment]
@levitte - updates pushed following @kaduk's review. Could you reconfirm?

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**